src/transformStateWithClones.js

```diff
@@ -7,7 +7,27 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => delete stateCopy[key]);
+        break;
+
+      case 'clear':
+        stateCopy = {};
+        break;
+    }
```

Comment:
[CODE STYLE]: A `default` case should be included in the `switch` statement for error handling. It's a good practice to handle unexpected `action.type` values, even if you expect a limited set of them. For example, you could throw an error or simply continue without modifying the state if an unknown action type is encountered.